symengine/polys/uintpoly.cpp

```diff
@@ -1,7 +1,311 @@
 #include <symengine/polys/uintpoly.h>
+#include <symengine/fields.h>
 
 namespace SymEngine
 {
+void UIntDict::itrunc(const integer_class &mod)
+{
+    for (auto it = dict_.begin(); it != dict_.end();) {
+        mp_fdiv_r(it->second, it->second, mod);
+        if (it->second > mod / 2)
+            it->second -= mod;
+        if (it->second == 0_z) {
+            it = dict_.erase(it);
+        } else {
+            ++it;
+        }
+    }
+}
+
+void UIntDict::zz_hensel_step(const integer_class &m, const UIntDict &f,
+                              const UIntDict &g, const UIntDict &h,
+                              const UIntDict &s, const UIntDict &t,
+                              const Ptr<UIntDict> &G, const Ptr<UIntDict> &H,
+                              const Ptr<UIntDict> &S, const Ptr<UIntDict> &T)
+{
+    UIntDict q, r;
+    integer_class M;
+    mp_pow_ui(M, m, 2);
+
+    auto e = f - g * h;
+    e.itrunc(M);
+
+    zz_divide(h, s * e, outArg(q), outArg(r));
+
+    q.itrunc(M);
+    r.itrunc(M);
+
+    UIntDict u = t * e + q * g;
+    *G = (g + u);
+    (*G).itrunc(M);
+    *H = (h + r);
+    (*H).itrunc(M);
+
+    u = s * (*G) + t * (*H);
+    auto b = (u - 1);
+    b.itrunc(M);
+
+    zz_divide(*H, s * b, outArg(q), outArg(r));
+    q.itrunc(M);
+    r.itrunc(M);
+
+    u = t * b + q * (*G);
+    *S = (s - r);
+    (*S).itrunc(M);
+    *T = (t - u);
+    (*T).itrunc(M);
+}
+
+integer_class UIntDict::l1_norm() const
+{
+    integer_class out(0_z);
+    for (auto &a : dict_) {
+        out += mp_abs(a.second);
+    }
+    return out;
+}
+
+UIntDict UIntDict::primitive() const
+{
+    integer_class gcd(0_z);
+    UIntDict out(dict_);
+    for (auto &a : dict_) {
+        mp_gcd(gcd, gcd, mp_abs(a.second));
+    }
+    for (auto &a : out.dict_) {
+        a.second /= gcd;
+    }
+    return out;
+}
+
+std::vector<UIntDict>
+UIntDict::zz_hensel_lift(const integer_class &p,
```

Comment:
Yes.
And I had read this algo from a source on internet, but after reading it from `Modern Computer Algebra` I am sure, I can make some more tweaks.